### DISCUSSION TITLE
How can I get url query in custom app component

### DISCUSSION THREAD

**Discussion by zerofront** [vercel/next.js#19003] [Help]
I want to use some global data in Next.js pages.

I do it with a custom App component and context API

in _app.js
```javascript
render() {
    const { Component, pageProps } = this.props;

    return (
      <CommonContext.Provider initialData={xxxxxxxxxx}>
        <Component {...pageProps} />
      </CommonContext.Provider>
    );
  }
```

**I want to use url query params as initial data, How can I get url query in custom app component ?**

**Comment by neg4n** [comment]
Hey @zerofront
It can be achieved using `withRouter()` function from `next/router`.
An example `_app.js` using `withRouter()` will look like this:
```jsx
import App from "next/app";
import { withRouter } from "next/router";

class MyApp extends App {
  render() {
    const { Component, pageProps, router } = this.props;
    const { test } = router.query;
    
    return (
      <>
        <h1>{test}</h1>
        <Component {...pageProps} />
      </>
    );
  }
}

export default withRouter(MyApp);
```
![image](https://user-images.githubusercontent.com/57688858/98634317-63324b00-2323-11eb-84bd-763466437a6d.png)

**Reply by zerofront** [reply]
@neg4n in server side, router.query return `{ amp: undefined }`

**Reply by neg4n** [reply]
@zerofront You have to check if router got initialized (can be done with `useEffect` or `componentDidMount` in your case) or you can do hacky workaround by installing [query-string](https://www.npmjs.com/package/query-string) and manually parsing `router.asPath` like this:
```js
const { queryParam } = queryString.parse(router.asPath.substring(1));
```

**Reply by zerofront** [reply]
@neg4n in server side _app.js, `router.asPath.substring(1)` doesn't return query string, just return route

```
ServerRouter {
  route: '/testpage',
  pathname: '/testpage',
  query: { amp: undefined },
  asPath: '/testpage',
  isFallback: false,
}
